Firestore/core/src/firebase/firestore/nanopb/message.h

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_NANOPB_MESSAGE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_NANOPB_MESSAGE_H_
+
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/nanopb/byte_string.h"
+#include "Firestore/core/src/firebase/firestore/nanopb/reader.h"
+#include "Firestore/core/src/firebase/firestore/nanopb/writer.h"
+#include "Firestore/core/src/firebase/firestore/remote/serializer.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "grpcpp/support/byte_buffer.h"
+
+namespace firebase {
+namespace firestore {
+
+namespace remote {
+namespace bridge {
+
+class DatastoreSerializer;
+class WatchStreamSerializer;
+class WriteStreamSerializer;
+
+}  // namespace bridge
+}  // namespace remote
+
+namespace nanopb {
+
+template <typename T>
+class Message;
+
+template <typename T>
+using MaybeMessage = util::StatusOr<Message<T>>;
```

Comment:
After some thinking, I have decided against wrapping the `StatusOr` inside `Message`, because this imposes checks even in cases where the creation of `Message` cannot fail.